sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_local.asciidoc

```diff
@@ -0,0 +1,312 @@
+
+The `work_group_local` class template acts as a view of an
+implementation-managed pointer to work-group local memory.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename T>
+class work_group_local {
+public:
+
+  work_group_local() = default;
+  work_group_local(const work_group_local&) = delete;
+  work_group_local& operator=(const work_group_local&) = delete;
+
+  operator T&() const noexcept;
+
+  // Available only if: std::is_array_v<T> == false
+  const work_group_local& operator=(const T& value) const noexcept;
+
+  T* operator&() const noexcept;
+
+private:
+  T* ptr; // exposition only
+
+};
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+`T` must be trivially constructible and trivially destructible.
+
+The storage for the object is allocated in work-group local memory before
+calling the user's kernel lambda, and deallocated when all work-items
+in the group have completed execution of the kernel.
+
+SYCL implementations conforming to the full feature set treat
+`work_group_local` similarly to the `thread_local` keyword, and when
+a `work_group_local` object is declared at block scope it behaves
+as if the `static` keyword was specified implicitly. SYCL implementations
+conforming to the reduced feature set require the `static` keyword to be
+specified explicitly.
+
+[NOTE]
+====
+If a `work_group_local` object is declared at function scope, the work-group
+local memory associated with the object will be identical for all usages of
+that function within the kernel. In cases where a function is called multiple
+times, developers must take care to avoid race conditions (e.g., by calling
+`group_barrier` before and after using the memory).
+====
+
+SYCL 2020 requires that all global variables accessed by a device function are
+`const` or `constexpr`. This extension lifts that restriction for
+`work_group_local` variables.
+
+[NOTE]
+====
+Since `work_group_local` acts as a view, wrapping an underlying pointer, a
+developer may still choose to declare variables as `const`.
+====
+
+When `T` is a class type or bounded array, the size of the allocation is known
+at compile-time, and a SYCL implementation may embed the size of the allocation
+directly within a kernel. Each instance of `work_group_local<T>` is associated
+with a unique allocation in work-group local memory.
+
+When `T` is an unbounded array, the size of the allocation is unknown at
+compile-time, and must be communicated to the SYCL implementation via the
+`work_group_local_memory_size` property. Every instance of
+`work_group_local<T>` is associated with a single, shared, allocation in
+work-group local memory.
+
+If the total amount of local memory requested (i.e., the sum of all memory
+requested by `local_accessor`, `group_local_memory`,
+`group_local_memory_for_overwrite` and `work_group_local`) exceeds a device's
+local memory capacity (as reported by `local_mem_size`) then the implementation
+must issue a diagnostic.
```

Comment:
Saying "issue a diagnostic" sounds like a compile-time error, but this is a runtime exception.  Should we say instead:

> ... then the implementation must throw a synchronous `exception` with the `errc::memory_allocation` error code from the kernel invocation command (e.g. `parallel_for`).

Reply:
Fixed in 7dd4d86fe2067c46beb313b044cdc31b0073bdad.